### PR TITLE
testing/debug: Enhancement for watch address by cmd line

### PR DESCRIPTION
**Pull request by no1wudi**
## Summary
This patch add some command line arguments for debug program, then you can use it to insert watchpoint or breakpoint to arbitrary address at runtime.
## Impact
New debug feature
## Testing
ESP32C3

### CODE REVIEW

File: testing/debug/debug.c [hunk 1]
```diff
@@ -39,6 +40,18 @@
...
+/****************************************************************************
+ * Name: debug_option
+ ****************************************************************************/
+
+struct debug_option
+{
+  int type;
+  int width;
+  bool cancel;
+  void *addr;
```
Comment (pkarashchenko) [inline testing/debug/debug.c:52]: FAR

File: testing/debug/debug.c [hunk 2]
```diff
@@ -236,6 +263,182 @@ nooptimiziation_function static void test_data(uint8_t *addr,
...
+ *   argv - An array of argument strings.
+ *   options - A pointer to a struct debug_option where the parsed options
+ *             will be stored.
+ *
+ * Returns:
+ *   true if the options were parsed successfully, false otherwise.
+ *
+ ****************************************************************************/
+
+static bool parse_options(int argc, char *argv[], struct debug_option *opt)
```
Comment (pkarashchenko) [inline testing/debug/debug.c:286]: FAR

File: testing/debug/debug.c [hunk 3]
```diff
@@ -236,6 +263,182 @@ nooptimiziation_function static void test_data(uint8_t *addr,
...
+
+            opt->type = GDB_STOPREASON_WATCHPOINT_RO;
+            opt->addr = (void *)(uintptr_t)strtoul(optarg, NULL, 0);
+            break;
+          case 'w':
+
+            /* Set a write watchpoint at the specified address */
+
+            opt->type = GDB_STOPREASON_WATCHPOINT_WO;
+            opt->addr = (void *)(uintptr_t)strtoul(optarg, NULL, 0);
```
Comment (pkarashchenko) [inline testing/debug/debug.c:305]: FAR

File: testing/debug/debug.c [hunk 4]
```diff
@@ -236,6 +263,182 @@ nooptimiziation_function static void test_data(uint8_t *addr,
...
+ * Parameters:
+ *   argc - The number of arguments passed to the program.
+ *   argv - An array of argument strings.
+ *
+ * Returns:
+ *   0 on success, or a negative error code on failure.
+ *
+ ****************************************************************************/
+
+static int handle_cmd(int argc, char *argv[])
```
Comment (pkarashchenko) [inline testing/debug/debug.c:385]: FAR

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
the test become useful after this patch, so it could move to apps/system.

**Comment by no1wudi** [comment]
> this test become useful after this patch, so it could move to apps/system.

OK

**Review by acassis — COMMENTED** [review]
Very nice @no1wudi !!!
I noticed there is not documentation to testing/debug: https://nuttx.apache.org/docs/latest/applications/testing/index.html could you please add it?

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**